pageserver/src/storage_sync.rs

```diff
@@ -883,42 +931,92 @@ impl RemoteTimelineClient {
-        if retries > 0 {
-            info!(
-                "remote task {} completed successfully after {} retries",
-                task.op, retries
-            );
-        } else {
-            info!("remote task {} completed successfully", task.op);
-        }
-
-        // The task has completed succesfully. Remove it from the in-progress list.
-        {
-            let mut upload_queue = self.upload_queue.lock().unwrap();
-            let mut upload_queue = upload_queue.initialized_mut().expect(
-                "callers are responsible for ensuring this is only called on initialized queue",
-            );
-            upload_queue.inprogress_tasks.remove(&task.task_id);
+        };
 
-            match task.op {
-                UploadOp::UploadLayer(_, _) => {
-                    upload_queue.num_inprogress_layer_uploads -= 1;
-                }
-                UploadOp::UploadMetadata(_, lsn) => {
-                    upload_queue.num_inprogress_metadata_uploads -= 1;
-                    upload_queue.last_uploaded_consistent_lsn = lsn; // XXX monotonicity check?
-                }
-                UploadOp::Delete(_, _) => {
-                    upload_queue.num_inprogress_deletions -= 1;
+        match outcome {
+            Outcome::TaskSucceeded => {
+                let mut upload_queue = self.upload_queue.lock().unwrap();
+                let mut upload_queue = upload_queue.initialized_mut().expect(
+                    "callers are responsible for ensuring this is only called on initialized queue",
+                );
+                upload_queue.inprogress_tasks.remove(&task.task_id);
+
+                match task.op {
+                    UploadOp::UploadLayer(_, _) => {
+                        upload_queue.num_inprogress_layer_uploads -= 1;
+                    }
+                    UploadOp::UploadMetadata(_, lsn) => {
+                        upload_queue.num_inprogress_metadata_uploads -= 1;
+                        upload_queue.last_uploaded_consistent_lsn = lsn; // XXX monotonicity check?
+                    }
+                    UploadOp::Delete(_, _) => {
+                        upload_queue.num_inprogress_deletions -= 1;
+                    }
+                    UploadOp::Barrier(_) => unreachable!(
+                        "Barrier operations are handled synchronously in launch_queued_tasks"
+                    ),
+                };
+
+                // Launch any queued tasks that were unblocked by this one.
+                self.launch_queued_tasks(upload_queue);
+            }
+            Outcome::ShutdownRequested => {
+                // Whichever *tasks* for this RemoteTimelineClient grabs the mutex first will transition the queue
+                // into shut down state, thereby dropping all off the queued *ops* which haven't become *tasks* yet.
+                // The other *tasks* will come here and observe an already shut down queue and hence simply wrap up their business.
```

Comment:
What exactly is the "wrap up their business" here? ISTM that they might well finish the upload successfully, breaking the earlier loop with Outcome::TaskSucceeded. Then they will call `upload_queue.initialized_mut()` above, and panic because the queue was already in ShutDown stat.

Reply:
Right, need a similar `match` in the `TaskSucceeded` case

Reply:
Will address this tomorrow.

Reply:
Done